### PR TITLE
Koa [#1425] and Node versioning (nvm)

### PR DESCRIPTION
**Pull request by ebramanti-techempower**
I've implemented the Koa framework as discussed in #1425. However, this brought forth a bigger problem: A lack of versioning for Node.js (depending on framework). Koa has a requirement of `>=v0.11` (needs ES6 Harmony features) for Node.js, however incrementing Node manually broke all the other frameworks.

The most ideal thing is to update each and every framework to work with the current latest version of Node. With `nvm` now used in the Node.js frameworks, this provides a way for quick Node versioning. Creating isolated environments for each framework is something that has been discussed in #995, and this is an important step forward for good versioning amongst the Node frameworks.

I'm going to continue work on getting these frameworks updated to `0.12.x`. In the meantime, I wanted to push the updated versioning so people can continue working on new frameworks (believe @zane-techempower is about to start work on Sails.js).

### CODE REVIEW

**Comment by ebramanti-techempower** [comment]
Need to do some more cleanup.

File: frameworks/JavaScript/hapi/app.js [hunk 1]
```diff
@@ -8,8 +8,8 @@ var cluster = require('cluster'),
 	Hapi = require('hapi'),
 	mongoose = require('mongoose'),
 	async = require('async'),
-	conn = mongoose.connect('mongodb://localhost/hello_world'),
-	connMap = { user: 'benchmarkdbuser', password: 'benchmarkdbpass', database: 'hello_world', host: 'localhost', charset: 'utf8' };
+	conn = mongoose.connect('mongodb://127.0.0.1/hello_world'),
+	connMap = { user: 'benchmarkdbuser', password: 'benchmarkdbpass', database: 'hello_world', host: '127.0.0.1', charset: 'utf8' };
```
Comment (LadyMozzarella) [inline frameworks/JavaScript/hapi/app.js]: Think you accidentally committed the 127.0.0.1 change. Switch back to localhost.

File: frameworks/JavaScript/koa/bash_profile.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+export NODE_HOME=${IROOT}/node-v0.12.0-linux-x64
```
Comment (LadyMozzarella) [inline frameworks/JavaScript/koa/bash_profile.sh]: We've removed the need for bash_profile.sh from TFB. If you could rebase and move NODE_HOME to setup that would be helpful.

File: frameworks/JavaScript/nodejs/hello.js [hunk 3]
```diff
@@ -19,9 +19,9 @@ var http = require('http')
   , url = require('url')
   , async = require('async')
   , mongoose = require('mongoose')
-  , conn = mongoose.connect('mongodb://localhost/hello_world')
+  , conn = mongoose.connect('mongodb://127.0.0.1/hello_world')
```
Comment (LadyMozzarella) [inline frameworks/JavaScript/nodejs/hello.js]: This file looks like `localhost` was replaced with `127.0.0.1` accidentally as well.

File: frameworks/JavaScript/koa/setup.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# ${NODE_HOME}/bin/node --harmony app.js &
+
+# !DO NOT REMOVE!
+#
+# It takes `node app` a few seconds to turn on and 
+# then fork. If you remove this sleep, the parent shell 
+# executing this script will be terminated before the 
+# application has time to awaken and be forked, and 
+# express will fail to be started
+sleep 5
```
Comment (LadyMozzarella) [inline frameworks/JavaScript/koa/setup.sh]: I believe we're already sleeping 60 seconds before starting each framework. Is this necessary?

File: frameworks/JavaScript/koa/setup.sh [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+# run app
+npm install
+node --harmony app &
+
+# export NODE_HOME=${IROOT}/node-v0.12.0-linux-x64
+# export PATH=$PATH:$NODE_HOME/bin
+
+# ${NODE_HOME}/bin/npm install
+# ${NODE_HOME}/bin/node --harmony app.js &
```
Comment (LadyMozzarella) [inline frameworks/JavaScript/koa/setup.sh]: Are all of these left commented out for a reason?

File: frameworks/JavaScript/koa/setup.sh (hunk 4 above)
Comment (msmith-techempower) [inline frameworks/JavaScript/koa/setup.sh]: Pinging @hamiltont
I DO believe this is necessary, but I don't fully recall why.

Comment (hamiltont) [inline frameworks/JavaScript/koa/setup.sh]: Now that #1505 is merged, this is no longer necessary. Explanation: 

The critical change was switching from `subprocess.check_call` to `subprocess.Popen` - the former will return when the subprocess is finished, while the latter will not. Without this sleep command, `setup.sh` would occasionally exit before the forking (e.g. `node --harmony app &`) was complete, python's subprocess would take care of cleaning up process resources, and therefore the app would never start. Now that we've switched to a non-terminating method of calling setup.sh from python, there is no need to sleep before returning.

File: frameworks/JavaScript/koa/setup.sh (hunk 5 above)
Comment (ebramanti-techempower) [inline frameworks/JavaScript/koa/setup.sh]: Ah, I missed a spot! Good catch!

File: frameworks/JavaScript/koa/setup.sh (hunk 4 above)
Comment (ebramanti-techempower) [inline frameworks/JavaScript/koa/setup.sh]: I'll get this update done for all the Node frameworks, seems that this is duplicated.

Comment (hamiltont) [inline frameworks/JavaScript/koa/setup.sh]: Thanks!

**Comment by ebramanti-techempower** [comment]
@msmith-techempower Do you want me to switch the `fw_depends` to `nvm` to be like `rvm.sh`, or is it OK how it is? Think we were talking about that over [here](https://github.com/TechEmpower/FrameworkBenchmarks/issues/1425#issuecomment-91645618).

**Comment by ebramanti-techempower** [comment]
I've updated the pull request with the requested changes!

**Comment by ebramanti-techempower** [comment]
So it looks like Express is failing on Travis, but this seems to be a module problem (not a problem with the Node version). Here is what I see in the log:

    Setup express: npm ERR! Error: No compatible version found: brace-expansion@'^1.0.0'

I'll fix this issue, as well as update the other Node frameworks, after this PR is merged to `master`.

**Comment by ebramanti-techempower** [comment]
Last commit should fix the Express issue.

**Comment by hamiltont** [comment]
> Last commit should fix the Express issue

Awesome, let's see what it says. For what it's worth, express on Travis-CI seems to succeed about 50% of the time. It's one of the ones that is not deterministic for some reason. My (total noob to npm) guess is that the dependencies are specified without a fixed version number e.g. brace-expansion>1.0 is ok vs a fixed e.g. brace-expansion 1.0.3.5a is required. I've seen something similar with python frameworks use a `requiremetns.txt` file without first issuing a `pip freeze` command to hardcode the requirement numbers

File: frameworks/JavaScript/nodejs/hello.js [hunk 6]
```diff
@@ -29,12 +29,12 @@ if (!windows) {
     
     Mapper.connect(connMap, {verbose: false, strict: false});
     var World = Mapper.map("World", "id", "randomNumber")
-    libmysql.connectSync('localhost', 'benchmarkdbuser', 'benchmarkdbpass', 'hello_world');
+    libmysql.connectSync('127.0.0.1', 'benchmarkdbuser', 'benchmarkdbpass', 'hello_world');
```
Comment (LadyMozzarella) [inline frameworks/JavaScript/nodejs/hello.js]: The `127.0.0.1`s snuck in again.

**Comment by ebramanti-techempower** [comment]
@hamiltont So the test passed for Express, I believe it is actually an issue with npm being out of date. What I now do is run `npm install -g npm` before running `npm install`, and this seems to have fixed the issue. Tests are all green, and I'm going to implement a fix for that line @LadyMozzarella caught.

**Comment by hamiltont** [comment]
> seems to have fixed the issue

Great, would love to see express start greenlighting continuously

**Comment by ebramanti-techempower** [comment]
@hamiltont Now my real work begins: updating Node tests to v0.12.x. I'll post this as an issue just so people know I'm working on it!

**Pull request merged**